pkg/controllers/resourceapply/resourceapply.go

```diff
@@ -158,6 +164,34 @@ func applyDeployment(ctx context.Context, client coreclientv1.Client, recorder r
 		return false, nil
 	}
 
+	// Check if deployment recreation needed
+	// Currently it is necessary if pod selector was changed
+	needRecreate := false
+	if !reflect.DeepEqual(existingCopy.Spec.Selector, required.Spec.Selector) {
+		needRecreate = true
+	}
+	if needRecreate {
+		recorder.Event(
+			existing, corev1.EventTypeNormal,
+			"Delete existing deployment", "Delete existing deployment to recreate it with new parameters",
+		)
+		// Perform dry run creation in order to validate deployment before deleting existing one
+		requiredCopy := required.DeepCopy()
+		requiredCopy.Name = fmt.Sprintf("%s-dry-run", requiredCopy.Name)
+		dryRunOpts := &coreclientv1.CreateOptions{DryRun: []string{metav1.DryRunAll}}
+		err = client.Create(ctx, requiredCopy, dryRunOpts)
+		if err != nil {
+			recorder.Event(existing, corev1.EventTypeWarning, "New resource validation failed", err.Error())
+			return false, err
+		}
+		err = client.Delete(ctx, existing)
+		if err != nil && !apierrors.IsNotFound(err) {
+			recorder.Event(existing, corev1.EventTypeWarning, "Deletion failed", err.Error())
+			return false, err
+		}
+		return applyDeployment(ctx, client, recorder, required)
```

Comment:
>Are we recursing on applyDeployment now?

Yes

>This seems like it could be dangerous, eg, what happens if the deployment doesn't go away immediately and takes some time to go away?

I expect it will be handled the same way as any other error returned from this function. Operator might go to degraded for a while, and reconciled then. https://github.com/openshift/cluster-cloud-controller-manager-operator/blob/ddc23ceec562563659d475a4a913a6d41b719d5b/pkg/controllers/clusteroperator_controller.go#L113

Reply:
Is there some way to either avoid this recursion or limit its depth? What are we achieving by recursing here? Is it possible maybe to break the logic into two smaller functions and call the exact logic we need rather than recursing?

Reply:
>Is there some way to either avoid this recursion or limit its depth?

Reply:
Ok, I think for now we can accept the recursion, but I want a comment on here that explains why the recursion is safe and why the recursion is required. I'd like to see similar comments on the Daemonset version

Reply:
removed recursion https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/174/commits/bec903d29c7b42c79afb801419a03b9fb5510654